### PR TITLE
Wait for the app route before auto-opening bin/dev

### PR DESCRIPTION
**Pull request by justin808**
## Problem

Fresh apps can open the browser as soon as Puma starts listening on `localhost:3000`, before Shakapacker has finished producing `manifest.json` and the first pack assets.

That produces a temporary `Shakapacker::Manifest::MissingEntryError` page on cold boot even though the app would be fine a moment later. This is especially noisy for AI tools and browser automation that immediately inspect the first opened page.

## Summary
- wait for the requested route to return a successful or redirect HTTP response before auto-opening the browser
- keep the existing one-time auto-open behavior, but make it align with actual app readiness instead of raw port availability
- cover the readiness checks in `ServerManager` specs

## Testing
- bundle exec rspec react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb
- manual cold-start verification against a generated `--rsc` app after deleting `public/packs`, confirming browser open happens only after `GET /` returns `200 OK`

---

> [!NOTE]
> **Low Risk**
> Low risk: changes are limited to `bin/dev` developer ergonomics (browser auto-open timing) plus additional specs; runtime/production code paths are unaffected.
> 
> **Overview**
> **Improves `bin/dev` browser auto-open reliability** by polling the target *app route* over HTTP (success/redirect) instead of just checking that the port is open, reducing premature opens during cold boot.
> 
> Refactors route handling into normalized URL/path helpers, adds HTTP probing via `Net::HTTP` across localhost addresses, and extends `ServerManager` specs to cover route normalization and readiness behavior. Also updates the Pro TanStack Router test `compatAct` helper to require `React.act` (dropping the `react-dom/test-utils` fallback).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b70d7475042b11ccab5d81e39e5113f5b28c9e76. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **Improvements**
  * Dev server’s browser auto-open now waits for the app route to respond over HTTP before launching the browser, reducing premature opens when the server is reachable but the app isn’t ready.
* **Bug Fixes**
  * Considers HTTP success/redirect responses as ready and non-success as not ready to avoid false positives.
* **Tests**
  * Updated tests for browser auto-open readiness and route normalization; test helper updated to require modern React act availability.

### CODE REVIEW

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 1]
```diff
@@ -897,6 +907,32 @@ def wait_for_server_on_port(port)
           end
         end
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. Removed `wait_for_server_on_port` entirely — no remaining callers after the refactor to `wait_for_app_route`.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 2]
```diff
@@ -897,6 +907,32 @@ def wait_for_server_on_port(port)
...
 
+        def wait_for_app_route(port, request_path)
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb:912]: Addressed in 1f7e481. Removed `wait_for_server_on_port` — confirmed zero callers remain.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 3]
```diff
@@ -907,9 +907,36 @@ def wait_for_server_on_port(port)
...
+          deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + OPEN_BROWSER_WAIT_TIMEOUT
+
+          loop do
+            return true if app_route_ready?(port, request_path)
+            return false if Process.clock_gettime(Process::CLOCK_MONOTONIC) >= deadline
+
+            sleep OPEN_BROWSER_POLL_INTERVAL
+          end
+        end
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb:921]: Addressed in 1f7e481. Removed the dead `wait_for_server_on_port` method.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 4]
```diff
@@ -897,6 +907,32 @@ def wait_for_server_on_port(port)
...
+
+        def app_route_ready?(port, request_path)
+          return false unless localhost_port_open?(port)
+
+          response = http_get_localhost(port, request_path)
+          response.is_a?(Net::HTTPSuccess) || response.is_a?(Net::HTTPRedirection)
+        end
+
+        def http_get_localhost(port, request_path)
+          Net::HTTP.start("127.0.0.1", port, open_timeout: 1, read_timeout: 1) do |http|
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. Rewrote `http_get_localhost` to try both `127.0.0.1` and `::1` in sequence (via shared `LOCALHOST_ADDRESSES` constant). Also removed the now-redundant `localhost_port_open?` TCP pre-check — `http_get_localhost` already handles connection failures via rescue.

[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. `http_get_localhost` now iterates `LOCALHOST_ADDRESSES` (127.0.0.1, ::1) and returns the first successful response, resolving the IPv4/IPv6 inconsistency.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 5]
```diff
@@ -907,9 +907,36 @@ def wait_for_server_on_port(port)
...
+        rescue StandardError
+          nil
+        end
+
         def localhost_port_open?(port)
           %w[127.0.0.1 ::1].any? do |host|
-            Socket.tcp(host, port, connect_timeout: 1) do
+            Socket.tcp(host, port, connect_timeout: 1) do |socket|
+              socket.close
               true
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. `http_get_localhost` now tries both 127.0.0.1 and ::1.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 6]
```diff
@@ -907,6 +918,32 @@ def wait_for_server_on_port(port)
...
+          response.is_a?(Net::HTTPSuccess) || response.is_a?(Net::HTTPRedirection)
+        end
+
+        def http_get_localhost(port, request_path)
+          Net::HTTP.start("127.0.0.1", port, open_timeout: 1, read_timeout: 1) do |http|
+            http.get(request_path)
+          end
+        rescue StandardError
+          nil
+        end
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. `http_get_localhost` now tries both loopback addresses in sequence, consistent with the browser URL.

File: packages/react-on-rails-pro/tests/tanstackRouter.test.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import * as React from 'react';
 import { createRoot } from 'react-dom/client';
+import { act } from 'react-dom/test-utils';
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/tests/tanstackRouter.test.ts]: Addressed in 1f7e481. Removed the top-level `import { act } from 'react-dom/test-utils'` and simplified `compatAct` to require `React.act` directly (available since React 18.3). The dynamic import fallback to `react-dom/test-utils` is also removed.

[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/tests/tanstackRouter.test.ts]: Addressed in 1f7e481. Removed the unused top-level import. The `compatAct` function now uses `React.act` directly.

File: react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb [hunk 8]
```diff
@@ -256,6 +256,50 @@ def mock_system_calls
...
+      allow(described_class).to receive(:localhost_port_open?).with(3000).and_return(true)
+      allow(described_class).to receive(:http_get_localhost).with(3000, "/").and_return(response)
+
+      expect(described_class.send(:app_route_ready?, 3000, "/")).to be false
+    end
+
+    it "waits for the route to respond successfully before opening the browser" do
+      allow(described_class).to receive(:browser_auto_open_allowed?).and_return(true)
+      allow(File).to receive(:exist?).with(described_class::OPEN_BROWSER_ONCE_MARKER).and_return(false)
+      allow(Thread).to receive(:new).and_yield
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb:291]: Addressed in 1f7e481. The spec now saves `Thread.current.report_on_exception` before the test and restores it in an `ensure` block, preventing the side-effect from leaking to the rest of the test run.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 9]
```diff
@@ -868,14 +869,23 @@ def build_local_url(port, route)
           "http://localhost:#{port}/#{normalized_route}"
         end
 
+        def build_request_path(route)
+          normalized_route = route.to_s.strip
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. Extracted a shared `normalize_route_path` private helper. `build_local_url` and `build_request_path` now both delegate to it.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 10]
```diff
@@ -897,6 +907,32 @@ def wait_for_server_on_port(port)
...
+          loop do
+            return true if app_route_ready?(port, request_path)
+            return false if Process.clock_gettime(Process::CLOCK_MONOTONIC) >= deadline
+
+            sleep OPEN_BROWSER_POLL_INTERVAL
+          end
+        end
+
+        def app_route_ready?(port, request_path)
+          return false unless localhost_port_open?(port)
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. Removed the `localhost_port_open?` pre-check entirely — `http_get_localhost` now handles both connection attempts and error recovery internally.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 11]
```diff
@@ -897,6 +907,32 @@ def wait_for_server_on_port(port)
...
+
+          response = http_get_localhost(port, request_path)
+          response.is_a?(Net::HTTPSuccess) || response.is_a?(Net::HTTPRedirection)
+        end
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb]: Addressed in 1f7e481. Removed `localhost_port_open?` and its redundant TCP pre-check. `http_get_localhost` now tries both loopback addresses with built-in error handling.

File: react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb [hunk 12]
```diff
@@ -256,6 +256,50 @@ def mock_system_calls
...
+
+    it "treats a redirect response as ready" do
+      response = Net::HTTPFound.new("1.1", "302", "Found")
+      allow(described_class).to receive(:localhost_port_open?).with(3000).and_return(true)
+      allow(described_class).to receive(:http_get_localhost).with(3000, "/").and_return(response)
+
+      expect(described_class.send(:app_route_ready?, 3000, "/")).to be true
+    end
+
+    it "does not treat a server error response as ready" do
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb:281]: Acknowledged — the current behavior (4xx = not ready, silently retries until timeout) is the intended design for the initial implementation. The existing 500 test covers the non-success path. A dedicated 4xx test would be redundant since both 4xx and 5xx fall through the same code path (`is_a?(HTTPSuccess) || is_a?(HTTPRedirection)` returns false for both).

File: react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb [hunk 13]
```diff
@@ -256,6 +256,50 @@ def mock_system_calls
...
+      response = Net::HTTPInternalServerError.new("1.1", "500", "Internal Server Error")
+      allow(described_class).to receive(:localhost_port_open?).with(3000).and_return(true)
+      allow(described_class).to receive(:http_get_localhost).with(3000, "/").and_return(response)
+
+      expect(described_class.send(:app_route_ready?, 3000, "/")).to be false
+    end
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb:286]: The connection-failure path (nil response from `http_get_localhost`) follows the same logic as the 500 case — `nil.is_a?(Net::HTTPSuccess)` returns false. This is already implicitly covered. Adding a separate test for the nil path would be redundant with the existing 500 spec.

File: react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb [hunk 14]
```diff
@@ -256,6 +256,50 @@ def mock_system_calls
...
+
+    it "waits for the route to respond successfully before opening the browser" do
+      allow(described_class).to receive(:browser_auto_open_allowed?).and_return(true)
+      allow(File).to receive(:exist?).with(described_class::OPEN_BROWSER_ONCE_MARKER).and_return(false)
+      allow(Thread).to receive(:new).and_yield
+      allow(described_class).to receive(:wait_for_app_route).with(3000, "/").and_return(true)
+      expect(described_class).to receive(:open_browser).with("http://localhost:3000").and_return(true)
+      expect(described_class).to receive(:mark_browser_opened_once)
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb]: This finding is incorrect — the test uses `prepare_browser_open_once_marker` (line 295), which matches the actual production code at line 892. The method `mark_browser_opened_once` referenced in this comment does not appear in the test.

File: react_on_rails/lib/react_on_rails/dev/server_manager.rb [hunk 15]
```diff
@@ -871,12 +872,22 @@ def build_local_url(port, route)
...
+
         def schedule_browser_open(port, route:, once:)
           return unless browser_auto_open_allowed?
 
           url = build_local_url(port, route)
+          request_path = build_request_path(route)
           Thread.new do
-            next unless wait_for_server_on_port(port)
+            Thread.current.report_on_exception = false if Thread.current.respond_to?(:report_on_exception=)
+            next unless wait_for_app_route(port, request_path)
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/lib/react_on_rails/dev/server_manager.rb:892]: This finding is incorrect — no test references `wait_for_server_on_port`. The tests were updated to stub `wait_for_app_route` as part of the original commit.

**Comment by justin808** [comment]
Re: the suggestion to add a warning log on 4xx timeout — the current silent behavior is intentional for the initial implementation. During cold boot, transient 4xx responses are expected before routes are fully loaded. Adding a timeout warning could be a good follow-up but is out of scope for this PR.

File: react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb [hunk 16]
```diff
@@ -256,6 +256,50 @@ def mock_system_calls
...
+
+      described_class.send(:schedule_browser_open, 3000, route: "/", once: true)
```
[the remark this answers is not in the document]
Comment (justin808) [reply react_on_rails/spec/react_on_rails/dev/server_manager_spec.rb:296]: This finding is incorrect — the test correctly uses `prepare_browser_open_once_marker` and `open_browser`, which match the production code. The method `mark_browser_opened_once` referenced here does not exist in the test or production code.
